### PR TITLE
[ fix #2858 ] Interleaving mutually-defined functions and datatypes

### LINKED ISSUE #2858: Feature request: Interleaving mutually-defined functions & datatypes
It's quite common in papers to see proofs of multiple statements by mutual induction. And it is also quite common to see all of the base cases dealt with in one go, and then all of the similar induction cases too. It'd be nice if Agda allowed this kind of didactic presentation. E.g. in normalisation by evaluation:

```agda
reify   : (σ : Type) → [ Val σ ⟶ Lam σ ]
reflect : (σ : Type) → [ Lam σ ⟶ Val σ ]

-- base case: everything matches up!
reify   α = id
reflect α = id

-- arrow case: eta expansion of lambdas!
reify   (σ ⇒ τ) = λ b → L (reify τ (b E.extend (reflect σ (V z))))
reflect (σ ⇒ τ) = λ b ρ v → reflect τ (A (ren ρ b) (reify σ v))
```

Similarly for inductive types:

```agda
data Even : Nat → Set
data Odd  : Nat → Set

data Even | Odd where

  -- base cases
  zero : Even 0
  one  : Odd 1

  -- step cases: the predecessor is of the opposite Even-ness
  succ : ∀ {n} → Odd n → Even (suc n)
  succ : ∀ {n} → Even n → Odd (suc n)
```

This could apply to IR too, I suppose. Although it's not as clear to me whether this is really readable (it would potentially lead to a limited version of the kind of solutions SHE was offering to the expression problem though):

```agda
data FreshList : Set
fresh : ℕ -> FreshList -> Set

[] : FreshList
fresh n [] = ⊤

_∷_by_ : ∀ n ns → fresh n ns → FreshList
fresh n (m ∷ ns by _) with m Nat.≟ n
... | yes p = ⊥
... | no ¬p = fresh n ns
```

### PR DESCRIPTION
**Pull request by gallais**
[ fix #2858 ] Interleaving mutually-defined functions and datatypes 

Prototypical example: a universe closed under `Nat` and `Sigma`
defined by IR. The interleaving mutual block allows us to consider
each constructor & its semantics in turn.

```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Sigma

interleaved mutual

  data Code : Set
  ⟦_⟧ : Code → Set

  constructor `Nat : Code
  ⟦ `Nat ⟧ = Nat

  constructor `Σ : (r : Code) → (⟦ r ⟧ → Code) → Code
  ⟦ `Σ A B ⟧ = Σ ⟦ A ⟧ λ a → ⟦ B a ⟧
```

### CODE REVIEW

**Comment by gallais** [comment]
I am getting some internal errors during interactive development.
Looks like I'll need to turn more `__IMPOSSIBLE__` into warnings.

Edit: fixed the lone signature issue. Still failing hard when there's
something constructor-shaped in a constructor block but its return
type does not make sense to us (e.g. it's a hole).

File: src/full/Agda/Interaction/Options/Warnings.hs [hunk 1]
```diff
@@ -306,9 +311,12 @@ warningNameDescription = \case
   EmptyRewritePragma_              -> "Empty `REWRITE' pragmas."
   EmptyWhere_                      -> "Empty `where' blocks."
   InvalidCatchallPragma_           -> "`CATCHALL' pragmas before a non-function clause."
+  InvalidConstructor_              -> "`constructor' blocks may only contain type signatures for constructors."
+  InvalidConstructorBlock_         -> "No `constructor' blocks outside of `infix mutual' blocks."
```
Comment (andreasabel) [inline src/full/Agda/Interaction/Options/Warnings.hs]: `infix mutual` is only a placeholder for a keyword determined by bikeshedding, I suppose?

**Review by andreasabel — CHANGES_REQUESTED** [review]
This is the 3rd flavor of mutual blocks, maybe we should just number the different types:
- `mutual`: old-style mutual blocks
- `mutual-i`: new-style mutual blocks.  No keyword so far, but it has been suggested to add a keyword for these, which would be optional (at least for a long deprecation phase). 
- `mutual-ii`: This "`infix mutual`" variant.
- ... (the next even more powerful variant)

Maybe arabic `mutual1`, `mutual2` or their subscript variant looks better.

File: src/full/Agda/Syntax/Concrete/Definitions.hs [hunk 2]
```diff
@@ -1513,8 +776,189 @@ niceDeclarations fixs ds = do
...
 
+    -- | Turn an new style interleaving mutual block into a new style mutual block
+    --   by grouping the declarations in blocks.
+    mkInfixMutual
+      :: Range                -- ^ Range of the whole @mutual@ block.
+      -> [NiceDeclaration]    -- ^ Declarations inside the block.
+      -> Nice NiceDeclaration -- ^ Returns a 'NiceMutual'.
+    mkInfixMutual r ds' = do
+      (other, (m, checks, _)) <- runStateT (groupByBlocks r ds') (empty, mempty, 0)
+      let dec k (InfixData (i, d@(NiceDataSig _ acc abs pc uc _ pars _)) ds) =
```
Comment (gallais) [inline src/full/Agda/Syntax/Concrete/Definitions.hs]: TODO: move these transformations next to the `InfixData` / `InfixFun` introduction site.

**Review by andreasabel — APPROVED** [review]
Not sure it makes sense for me to try a proper review...  It is a new feature, so it will be tried in practice.

Where did the discussion take place? (Link!)

**Comment by gallais** [comment]
Discussion took place on Zoom during the last wrap up meeting & was
summarised on zulip: https://agda.zulipchat.com/#narrow/stream/257010-aim-xxxiii/topic/Bikeshedding.20.235001

**Pull request merged**